### PR TITLE
Add index as an optional parameter to #each blocks

### PR DESCRIPTION
**Pull request by tim-evans**
This is the work for #10158.

### CODE REVIEW

**Comment by mmun** [comment]
Can you add

```hbs
run(function() {
  view.controller.unshiftObject({ name: "Bob" });
});
equal(view.$().text(), "0. Bob1. Adam2. Steve");
```

**Comment by mmun** [comment]
Totally agree that this test belongs in ember-views, but it's fine to leave it in ember-htmlbars for now. 

Historically the view tests have been in the templating packages because its easier to setup the view hierarchy. I do plan on moving the more unit-y tests to ember-views one day.

**Comment by tim-evans** [comment]
Yes! I'll do that now

**Comment by tim-evans** [comment]
Should I push with that test failing? I don't really know how to fix that

**Comment by mmun** [comment]
@tim-evans Ah, turns out contentIndex doesn't update currently. It will be a bit more complicated.

**Comment by mmun** [comment]
@tim-evans I got it working. See if you can incorporate this: https://github.com/mmun/ember.js/tree/each-index

**Comment by mmun** [comment]
Can you add a feature flag and squash all the commits? Otherwise its good to merge.

**Comment by ebryn** [comment]
:+1: great work guys

**Comment by tim-evans** [comment]
@mmun done

File: packages/ember-views/lib/views/collection_view.js [hunk 1]
```diff
@@ -357,13 +357,35 @@ var CollectionView = ContainerView.extend({
         item = content.objectAt(idx);
 
         itemViewProps.content = item;
-        itemViewProps._blockArguments = [item];
+        if (!Ember.FEATURES.isEnabled('ember-each-with-index')) {
```
Comment (rwjblue) [inline packages/ember-views/lib/views/collection_view.js]: Due to a limitation of `defeatureify` we cannot use `!Ember.FEATUREs.isEnabled`, you have to use:

```javascript
if (Ember.FEATURES.isEnabled('blah')) {

} else {
  /* your stuff here */
}
```

File: features.json [hunk 2]
```diff
@@ -13,7 +13,8 @@
     "ember-routing-transitioning-classes": true,
     "new-computed-syntax": null,
     "ember-testing-checkbox-helpers": null,
-    "ember-metal-stream": null
+    "ember-metal-stream": null,
+    "ember-each-with-index": null
```
Comment (rwjblue) [inline features.json]: The general pattern for feature flag names is `<package-name>-<short-feature-name>`.  In this case, I'd suggest `ember-htmlbars-each-with-index`.

File: packages/ember-views/lib/views/collection_view.js (hunk 1 above)
Comment (tim-evans) [inline packages/ember-views/lib/views/collection_view.js]: So I need a blank if for this one?

**Comment by tim-evans** [comment]
@rwjblue done.

File: packages/ember-views/lib/views/collection_view.js (hunk 1 above)
Comment (rwjblue) [inline packages/ember-views/lib/views/collection_view.js]: Sadly yes, I am sorry. :(

File: packages/ember-views/lib/views/collection_view.js [hunk 3]
```diff
@@ -357,13 +357,36 @@ var CollectionView = ContainerView.extend({
...
 
         view = this.createChildView(itemViewClass, itemViewProps);
 
+        if (Ember.FEATURES.isEnabled('ember-htmlbars-each-with-index')) {
+          if (this.blockParams > 1) {
+            view._blockArguments = [item, view.getStream('_view.contentIndex')];
+          } else if (this.blockParams === 1) {
+            view._blockArguments = [item];
+          }
+        }
```
Comment (mmun) [inline packages/ember-views/lib/views/collection_view.js]: You can move the original logic down here,

```js
if (Ember.FEATURES.isEnabled('ember-htmlbars-each-with-index')) {
  if (this.blockParams > 1) {
    view._blockArguments = [item, view.getStream('_view.contentIndex')];
  } else if (this.blockParams === 1) {
    view._blockArguments = [item];
  }
} else {
  if (this.blockParams > 0) {
    view._blockArguments = [item];
  }
}
```

File: packages/ember-views/lib/views/collection_view.js [hunk 4]
```diff
@@ -357,13 +357,36 @@ var CollectionView = ContainerView.extend({
         item = content.objectAt(idx);
 
         itemViewProps.content = item;
-        itemViewProps._blockArguments = [item];
+        if (Ember.FEATURES.isEnabled('ember-htmlbars-each-with-index')) {
+        } else {
+          itemViewProps._blockArguments = [item];
```
Comment (mmun) [inline packages/ember-views/lib/views/collection_view.js]: You can remove this and...

**Comment by tim-evans** [comment]
Oh goodness. Lots of tweaks. I hope this is the last one!

**Comment by tim-evans** [comment]
Thanks! :+1: (especially @mmun !)

**Comment by rwjblue** [comment]
@tim-evans - thank you!

**Comment by rwjblue** [comment]
As of https://github.com/emberjs/ember.js/pull/10193 this is enabled by default on canary (and slated for 1.11 barring any issues).

**Pull request merged**